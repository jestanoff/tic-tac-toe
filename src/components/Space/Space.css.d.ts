// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  blueSecondary: string;
  bottom: string;
  greyPrimary: string;
  left: string;
  right: string;
  space: string;
  symbolBlack: string;
  symbolWhite: string;
  top: string;
}
export const cssExports: CssExports;
export default cssExports;
