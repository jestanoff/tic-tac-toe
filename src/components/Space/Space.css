@value blueSecondary, greyPrimary from "../../css/colors.css";

.space {
  background: transparent;
  border: 3px solid blueSecondary;
  box-sizing: border-box;
  float: left;
  font-size: 60px;
  height: 80px;
  text-align: center;
  user-select: none;
  width: 33.33%;
}

.left {
  border-left-color: transparent;
}

.right {
  border-right-color: transparent;
}

.top {
  border-top-color: transparent;
}

.bottom {
  border-bottom-color: transparent;
}

.symbolWhite {
  color: white;
}

.symbolBlack {
  color: greyPrimary;
}
