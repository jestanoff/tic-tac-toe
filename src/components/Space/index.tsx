import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Mark from 'components/Mark';
import { CSS_CLASSES, PLAYER_X, SYMBOLS } from '../../constants';
import styles from './Space.css';

const cx = classNames.bind(styles);

const Space = ({ id, handleSpaceClick, status }) => {
  const hasBeenMarked = !!status;
  const title = `${CSS_CLASSES[id].join(' ')} ${hasBeenMarked ? `space marked by ${SYMBOLS[status]}` : 'empty space'}`;
  const capitalisedTitle = `${title.charAt(0).toUpperCase()}${title.slice(1)}`;
  const handleClick = React.useCallback(() => handleSpaceClick(id), [handleSpaceClick, id]);

  return (
    <button className={cx('space', CSS_CLASSES[id])} onClick={handleClick} title={capitalisedTitle}>
      {hasBeenMarked && <Mark type={status === PLAYER_X ? 'cross' : 'circle'} isAnimated />}
    </button>
  );
};

Space.propTypes = {
  handleSpaceClick: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  status: PropTypes.number,
};

Space.defaultProps = {
  status: 0,
};

export default React.memo(Space);
