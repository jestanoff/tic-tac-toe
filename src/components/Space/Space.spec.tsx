import * as React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Space from './index';

const emptyFn = () => undefined;

describe('<Space />', () => {
  test('should render a top space', () => {
    const { asFragment } = render(<Space id={0} handleSpaceClick={emptyFn} />);

    expect(screen.getByRole('button', { name: /^Top left empty space$/i })).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should render top left empty space', () => {
    render(<Space id={0} handleSpaceClick={emptyFn} />);
    expect(screen.getByRole('button', { name: /^Top left empty space$/i })).toHaveClass('space top left');
  });

  test('should render top left marked by cross space', () => {
    render(<Space id={0} handleSpaceClick={emptyFn} status={1} />);
    expect(screen.getByRole('button', { name: /^Top left space marked by cross$/i })).toHaveClass('space top left');
  });

  test('should render top left marked by circle space', () => {
    render(<Space id={0} handleSpaceClick={emptyFn} status={2} />);
    expect(screen.getByRole('button', { name: /^Top left space marked by circle$/i })).toHaveClass('space top left');
  });

  test('should render top center empty space', () => {
    render(<Space id={1} handleSpaceClick={emptyFn} />);
    expect(screen.getByRole('button', { name: /^Top center empty space$/i })).toHaveClass('space top center');
  });

  test('should render top center marked by cross space', () => {
    render(<Space id={1} handleSpaceClick={emptyFn} status={1} />);
    expect(screen.getByRole('button', { name: /^Top center space marked by cross$/i })).toHaveClass('space top center');
  });

  test('should render top center marked by circle space', () => {
    render(<Space id={1} handleSpaceClick={emptyFn} status={2} />);
    expect(screen.getByRole('button', { name: /^Top center space marked by circle$/i })).toHaveClass(
      'space top center',
    );
  });

  test('should render top right empty space', () => {
    render(<Space id={2} handleSpaceClick={emptyFn} />);
    expect(screen.getByRole('button', { name: /^Top right empty space$/i })).toHaveClass('space top right');
  });

  test('should render top right marked by cross space', () => {
    render(<Space id={2} handleSpaceClick={emptyFn} status={1} />);
    expect(screen.getByRole('button', { name: /^Top right space marked by cross$/i })).toHaveClass('space top right');
  });

  test('should render top right marked by circle space', () => {
    render(<Space id={2} handleSpaceClick={emptyFn} status={2} />);
    expect(screen.getByRole('button', { name: /^Top right space marked by circle$/i })).toHaveClass('space top right');
  });

  test('should render left middle empty space', () => {
    render(<Space id={3} handleSpaceClick={emptyFn} />);
    expect(screen.getByRole('button', { name: /^Left middle empty space$/i })).toHaveClass('space left middle');
  });

  test('should render left middle marked by cross space', () => {
    render(<Space id={3} handleSpaceClick={emptyFn} status={1} />);
    expect(screen.getByRole('button', { name: /^Left middle space marked by cross$/i })).toHaveClass(
      'space left middle',
    );
  });

  test('should render left middle marked by circle space', () => {
    render(<Space id={3} handleSpaceClick={emptyFn} status={2} />);
    expect(screen.getByRole('button', { name: /^Left middle space marked by circle$/i })).toHaveClass(
      'space left middle',
    );
  });

  test('should render center middle empty space', () => {
    render(<Space id={4} handleSpaceClick={emptyFn} />);
    expect(screen.getByRole('button', { name: /^Center middle empty space$/i })).toHaveClass('space center middle');
  });

  test('should render center middle marked by cross space', () => {
    render(<Space id={4} handleSpaceClick={emptyFn} status={1} />);
    expect(screen.getByRole('button', { name: /^Center middle space marked by cross$/i })).toHaveClass(
      'space center middle',
    );
  });

  test('should render center middle marked by circle space', () => {
    render(<Space id={4} handleSpaceClick={emptyFn} status={2} />);
    expect(screen.getByRole('button', { name: /^Center middle space marked by circle$/i })).toHaveClass(
      'space center middle',
    );
  });

  test('should execute handleSpaceClick function on click', async () => {
    const handleSpaceClickMock = jest.fn();
    render(<Space id={6} handleSpaceClick={handleSpaceClickMock} status={1} />);

    await userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(handleSpaceClickMock).toHaveBeenCalled());
    expect(handleSpaceClickMock).toHaveBeenCalledTimes(1);
    expect(handleSpaceClickMock).toHaveBeenCalledWith(6);
  });
});
