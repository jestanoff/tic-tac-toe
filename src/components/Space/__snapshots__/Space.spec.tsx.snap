// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Space /> should render a top space 1`] = `
<DocumentFragment>
  <button
    class="space top left"
    title="Top left empty space"
  />
</DocumentFragment>
`;
